image.go

```diff
@@ -1046,85 +1048,103 @@ func (rc *RegClient) ImageExport(ctx context.Context, r ref.Ref, outStream io.Wr
 		mode:  0644,
 	}
 
-	// retrieve image manifest
-	m, err := rc.ManifestGet(ctx, r)
-	if err != nil {
-		rc.slog.Warn("Failed to get manifest",
-			slog.String("ref", r.CommonName()),
-			slog.String("err", err.Error()))
-		return err
-	}
-
 	// build/write oci-layout
 	ociLayout := v1.ImageLayout{Version: ociLayoutVersion}
-	err = twd.tarWriteFileJSON(ociLayoutFilename, ociLayout)
+	err := twd.tarWriteFileJSON(ociLayoutFilename, ociLayout)
 	if err != nil {
 		return err
 	}
 
-	// create a manifest descriptor
-	mDesc := m.GetDescriptor()
-	if mDesc.Annotations == nil {
-		mDesc.Annotations = map[string]string{}
-	}
-	mDesc.Annotations[annotationImageName] = opt.exportRef.CommonName()
-	mDesc.Annotations[annotationRefName] = opt.exportRef.Tag
-
-	// generate/write an OCI index
+	// generate an OCI index
 	ociIndex.Versioned = v1.IndexSchemaVersion
-	ociIndex.Manifests = []descriptor.Descriptor{mDesc} // initialize with the descriptor to the manifest list
-	err = twd.tarWriteFileJSON(ociIndexFilename, ociIndex)
-	if err != nil {
-		return err
-	}
+	ociIndex.Manifests = []descriptor.Descriptor{} // initialize with the descriptor to the manifest list
+
+	// generate legacy docker manifests
+	dockerManifests := []dockerTarManifest{}
 
-	// append to docker manifest with tag, config filename, each layer filename, and layer descriptors
-	if mi, ok := m.(manifest.Imager); ok {
-		conf, err := mi.GetConfig()
+	for _, r := range refs {
+		// retrieve image manifest
+		m, err := rc.ManifestGet(ctx, r)
 		if err != nil {
+			rc.slog.Warn("Failed to get manifest",
+				slog.String("ref", r.CommonName()),
+				slog.String("err", err.Error()))
 			return err
 		}
-		if err = conf.Digest.Validate(); err != nil {
-			return err
-		}
-		refTag := opt.exportRef.ToReg()
-		if refTag.Digest != "" {
-			refTag.Digest = ""
-		}
-		if refTag.Tag == "" {
-			refTag.Tag = "latest"
-		}
-		dockerManifest := dockerTarManifest{
-			RepoTags:     []string{refTag.CommonName()},
-			Config:       tarOCILayoutDescPath(conf),
-			Layers:       []string{},
-			LayerSources: map[digest.Digest]descriptor.Descriptor{},
+
+		var er ref.Ref
+		if opt.exportRef.IsZero() {
+			// Clear the Digest
+			er = r.SetTag(r.Tag)
+		} else {
+			er = opt.exportRef
 		}
-		dl, err := mi.GetLayers()
-		if err != nil {
-			return err
+		// create a manifest descriptor
+		mDesc := m.GetDescriptor()
+		if mDesc.Annotations == nil {
+			mDesc.Annotations = map[string]string{}
 		}
-		for _, d := range dl {
-			if err = d.Digest.Validate(); err != nil {
+		mDesc.Annotations[annotationImageName] = er.CommonName()
+		mDesc.Annotations[annotationRefName] = er.Tag
```

Comment:
This has the same issue when the feature was previously requested that multiple images with the same tag (but different source repository) will get added with the same OCI annotation and cannot be differentiated in the OCI layout with the existing `ocidir` reference.

Reply:
`docker image save` also outputs index.json with same tag. These are valid OCI tars that one can come across from existing tools anyway.

```
$ docker image save {debian,alpine}:latest|tar xf - -O index.json|jq
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
      "digest": "sha256:e11072c1614c08bf88b543fcfe09d75a0426d90896408e926454e88078274fcb",
      "size": 1854,
      "annotations": {
        "containerd.io/distribution.source.docker.io": "library/debian",
        "io.containerd.image.name": "docker.io/library/debian:latest",
        "org.opencontainers.image.ref.name": "latest"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:029a752048e32e843bd6defe3841186fb8d19a28dae8ec287f433bb9d6d1ad85",
      "size": 1022,
      "annotations": {
        "io.containerd.image.name": "docker.io/library/alpine:latest",
        "org.opencontainers.image.ref.name": "latest"
      }
    }
  ]
}
```

Reply:
If the export contains an OCI Layout, and regclient and other tools work with an OCI Layout, then making sure that structure is usable and consistent is important to me. There are structural decisions made in treating the OCI Layout as a repository, and not a transport of content in another repository, which break when those assumptions are changed.

Docker is not OCI, it's a container engine that sometimes implements OCI specs. In this case, the spec is loosely defined causing incompatibilities between implementations.